File: generate_inserts.js
const fs = require('fs');

// Read app.js
const appCode = fs.readFileSync('C:\\Users\\User\\OneDrive\\Ai\\app.js', 'utf8');

// Extract SEED_TRANSACTIONS
const startToken = 'const SEED_TRANSACTIONS = [';
const endToken = '];';
const startIdx = appCode.indexOf(startToken);
if (startIdx === -1) {
  console.error("Could not find SEED_TRANSACTIONS in app.js");
  process.exit(1);
}

const slice = appCode.substring(startIdx + startToken.length - 1);
const endIdx = slice.indexOf(endToken);
if (endIdx === -1) {
  console.error("Could not find end of SEED_TRANSACTIONS");
  process.exit(1);
}

const arrayStr = slice.substring(0, endIdx + 1);
const seedTransactions = eval(arrayStr);

function parseCustomDate(dateStr) {
  if (!dateStr) return "";
  dateStr = dateStr.trim();
  
  if (/^\d{4}-\d{2}-\d{2}$/.test(dateStr)) {
    return dateStr;
  }
  
  const parts = dateStr.split(/[-/]/);
  if (parts.length !== 3) return dateStr;
  
  let p1 = parseInt(parts[0], 10);
  let p2 = parseInt(parts[1], 10);
  let yr = parseInt(parts[2], 10);
  if (yr < 100) yr += 2000;
  
  let day, month;
  if (p1 > 12) {
    day = p1;
    month = p2;
  } else if (p2 > 12) {
    day = p2;
    month = p1;
  } else {
    if (p2 === 5) {
      month = 5;
      day = p1;
    } else if (p1 === 5) {
      month = 5;
      day = p2;
    } else if (p1 === 6) {
      month = 6;
      day = p2;
    } else if (p2 === 6) {
      month = 6;
      day = p1;
    } else {
      day = p1;
      month = p2;
    }
  }
  
  const mm = String(month).padStart(2, "0");
  const dd = String(day).padStart(2, "0");
  return `${yr}-${mm}-${dd}`;
}

let sql = '';
seedTransactions.forEach(t => {
  const isoDate = parseCustomDate(t.date);
  const totalVal = parseFloat(t.total);
  const loc = t.location || '';
  const rem = t.remark || '';
  const updatedAt = Date.now(); // or a constant to keep it stable
  
  // Format SQL string literals
  const idStr = t.id.replace(/'/g, "''");
  const platformStr = t.platform.replace(/'/g, "''");
  const catStr = t.category.replace(/'/g, "''");
  const locStr = loc.replace(/'/g, "''");
  const remStr = rem.replace(/'/g, "''");
  
  sql += `INSERT OR IGNORE INTO transactions (id, date, type, platform, total, category, location, remark, updated_at) VALUES ('${idStr}', '${isoDate}', '${t.type}', '${platformStr}', ${totalVal}, '${catStr}', '${locStr}', '${remStr}', ${updatedAt});\n`;
});

fs.writeFileSync('C:\\Users\\User\\OneDrive\\Ai\\seed_data.sql', sql, 'utf8');
console.log(`Generated seed_data.sql with ${seedTransactions.length} records.`);

File: generate_inserts.py
import re
import json
import time

with open("C:/Users/User/OneDrive/Ai/app.js", "r", encoding="utf-8") as f:
    content = f.read()

# Extract SEED_TRANSACTIONS = [ ... ]
pattern = r"const SEED_TRANSACTIONS = \[(.*?)\];"
match = re.search(pattern, content, re.DOTALL)
if not match:
    print("Could not find SEED_TRANSACTIONS")
    exit(1)

# Format the matching text to valid JSON
list_content = match.group(1).strip()

# Let's clean the javascript array to be valid JSON
# We can parse the individual object literals using regex or simple parser
# Let's find all { ... } blocks
blocks = re.findall(r"\{(.*?)\}", list_content, re.DOTALL)

def parse_custom_date(date_str):
    date_str = date_str.strip()
    if re.match(r"^\d{4}-\d{2}-\d{2}$", date_str):
        return date_str
    
    parts = re.split(r"[-/]", date_str)
    if len(parts) != 3:
        return date_str
    
    p1 = int(parts[0])
    p2 = int(parts[1])
    yr = int(parts[2])
    if yr < 100:
        yr += 2000
    
    if p1 > 12:
        day = p1
        month = p2
    elif p2 > 12:
        day = p2
        month = p1
    else:
        if p2 == 5:
            month = 5
            day = p1
        elif p1 == 5:
            month = 5
            day = p2
        elif p1 == 6:
            month = 6
            day = p2
        elif p2 == 6:
            month = 6
            day = p1
        else:
            day = p1
            month = p2
            
    return f"{yr:04d}-{month:02d}-{day:02d}"

sql_lines = []
for b in blocks:
    # parse key-value pairs
    pairs = re.findall(r"(\w+):\s*(['\"].*?['\"]|[\d.]+)", b)
    t_data = {}
    for k, v in pairs:
        # Strip quotes if string
        val = v.strip()
        if (val.startswith("'") and val.endswith("'")) or (val.startswith('"') and val.endswith('"')):
            val = val[1:-1]
        else:
            try:
                val = float(val) if '.' in val else int(val)
            except ValueError:
                pass
        t_data[k] = val
        
    # Ensure optional fields exist
    txn_id = t_data.get("id", "")
    date_str = parse_custom_date(t_data.get("date", ""))
    txn_type = t_data.get("type", "")
    platform = t_data.get("platform", "")
    total = float(t_data.get("total", 0))
    category = t_data.get("category", "")
    location = t_data.get("location", "")
    remark = t_data.get("remark", "")
    updated_at = int(time.time() * 1000)
    
    # Escape quotes
    txn_id = txn_id.replace("'", "''")
    platform = platform.replace("'", "''")
    category = category.replace("'", "''")
    location = location.replace("'", "''")
    remark = remark.replace("'", "''")
    
    sql_lines.append(
        f"INSERT OR IGNORE INTO transactions (id, date, type, platform, total, category, location, remark, updated_at) "
        f"VALUES ('{txn_id}', '{date_str}', '{txn_type}', '{platform}', {total}, '{category}', '{location}', '{remark}', {updated_at});"
    )

with open("C:/Users/User/OneDrive/Ai/seed_data.sql", "w", encoding="utf-8") as out:
    out.write("\n".join(sql_lines))

print(f"Successfully generated seed_data.sql with {len(sql_lines)} queries.")
